### PR TITLE
Dependency graph validation

### LINKED ISSUE #4: List unmet needs in error message
Knowing which needs are unmet is helpful information when setting up or changing larger dependency graphs.

```
Cannot bootstrap: unable to provide for: expressApp; http
```

In my case might inform me that the unmet needs were: `config; controllers; health`

_(Excuse the nitpicking, commenting with some points that I run into porting an existing project to booture as I run into them.)_

### PR DESCRIPTION
**Pull request by Avaq**
To close #4 

/cc @beanow

### CODE REVIEW

**Comment by Beanow** [comment]
Giving this a read over, LGTM. Just minor suggestions to add.

In init-system I added an explicit validation for self-dependent definitions (i.e. config needs config), looks like this should be covered by circular dependency checks, but might warrant a test case so future changes wouldn't cause infinite loops or the likes.

I would rephrase `[x] needs [y], which has no provider` to `[y] has no providers, needed by [x]`. This way all error messages have the problem at the start of the sentence.

**Comment by Avaq** [comment]
Thank you for having a look @Beanow :)

> [self-dependency] might warrant a test case so future changes wouldn't cause infinite loops or the likes

It has one already: https://github.com/fluture-js/booture/pull/5/files#diff-6cde92536c50a2e90c7e5b6c9ee1666eR23

> I would rephrase `[x] needs [y], which has no provider` to `[y] has no providers, needed by [x]`

I tried it, but note how it reads when there's multiple bootstrappers depending on `y`:

```txt
[y] has no providers, needed by [x]
[y] has no providers, needed by [z]
```

vs

```txt
[x] needs [y], which has no provider
[z] needs [y], which has no provider
```

I find the latter reads more pleasantly.

> [With the former,] all error messages have the problem at the start of the sentence

It depends on what the problem is. The absence of a provider, or a typo in your needs.

**Comment by dicearr** [comment]
I do not see any strong reason for deferring the validation of the bootstrappers. The provided bootstrappers need to be a directed acyclic graph (with single parent). I would prefer throwing a TypeError if the argument is not matching the expected type than using the rejection branch.

**Comment by Beanow** [comment]
@dicearr throwing instead of rejecting makes sense to me as well. It's a definite implementation error versus something like failing to acquire a resource because a database couldn't be reached.

I'm not sure the current implementation of booture has/needs a single parent constraint in it's graph through. I find this single parent to be the callback you give to the `runHook` untagged function instead of one of the bootstrappers.

The single parent constraint exists in `init-system` and tends to create a pretty awkward bootstrapper at the root for daemons. One which does not acquire or dispose any resources, must be tailored to have a dependency on all the services you want initialized even though you don't actually use them as a dependency and listens for `SIGTERM` and `SIGINT` so as to become a daemon.

**Comment by dicearr** [comment]
@Beanow you're right. The current implementation requires a single provider per `need` which is not the same as a single parent.

**Comment by Avaq** [comment]
One of the benefits of Booture is that we can represent multiple independent DAGS, each of which are interpreted in parallel. Not ever going to change! :)

I do agree with the `throw` thing. A nice consequence of which is that we can make the type of `bootstrap` more generic (not specialized in Error).

**Comment by Avaq** [comment]
@Beanow @dicearr Changed to throwing the error instead. See the fixup commit.

File: index.mjs [hunk 1]
```diff
@@ -106,7 +106,56 @@ import {hookAll, acquire, Hook} from 'fluture-hooks';
 // hasProp :: Object ~> String -> Boolean
 const hasProp = Object.prototype.hasOwnProperty;
 
-// callBootstrappers :: (Array (Bootstrapper), StrMap Resource) -> Hook (StrMap Resource)
+const error = message => new TypeError(message);
+const ok = Hook.of(null);
```
Comment (dicearr) [inline index.mjs]: I think this two functions are not used anymore.

**Review by dicearr — CHANGES_REQUESTED** [review]

File: index.mjs (hunk 1 above)
Comment (Avaq) [reply index.mjs]: You are quite right.

**Review by dicearr — DISMISSED** [review]

**Comment by Avaq** [comment]
@dicearr Squashed and rebased.

**Review by dicearr — APPROVED** [review]
:+1:
